Runtime/Proc/ExecutionContext.cs

```diff
@@ -4,65 +4,105 @@ namespace Schema.Internal
+        public static ExecutionContext Current { get; set; }
+        public SchemaAgent Agent { get; }
 
-        public ExecutableNode node
+        public ExecutableNode Node
         {
             get => _node;
             set
             {
-                if (value == null)
+                if (value == null || value == _node)
                     return;
 
-                if (status == NodeStatus.Running || (_node != null && value.index - node.index < 0))
+                if (Status == NodeStatus.Running || (_node != null && value.Index - _node.Index < 0))
                 {
-                    lastStatus[_node.index] = status;
+                    _lastStatus[_node.Index] = Status;
 
-                    if (value.index == 0)
-                        lastStatus[0] = status;
+                    if (value.Index == 0)
+                        _lastStatus[0] = Status;
                 }
 
-                last = _node;
+                StepOut(_node);
+                StepIn(value);
+
+                Last = _node;
                 _node = value;
             }
         }
 
-        public ExecutableNode last { get; private set; }
+        public ExecutableNode Last { get; private set; }
 
-        public NodeStatus status { get; set; }
+        public NodeStatus Status { get; set; }
 
-        public bool forceActionConditionalEvaluation { get; set; }
+        public bool ForceActionConditionalEvaluation { get; set; }
 
         public void ForceStatus(ExecutableNode node)
         {
-            if (node.index > 0)
-                lastStatus[node.index] = status;
+            if (node.Index > 0)
+                _lastStatus[node.Index] = Status;
         }
 
         public void RemoveStatus(ExecutableNode node)
         {
-            if (node.index > 0 && lastStatus.ContainsKey(node.index))
-                lastStatus.Remove(node.index);
+            if (node.Index > 0 && _lastStatus.ContainsKey(node.Index))
+                _lastStatus.Remove(node.Index);
         }
 
         public NodeStatus? GetLastStatus(int index)
         {
-            if (!lastStatus.ContainsKey(index))
+            if (!_lastStatus.ContainsKey(index))
                 return null;
 
-            return lastStatus[index];
+            return _lastStatus[index];
+        }
+
+        private void StepIn(ExecutableNode node)
+        {
+            if (!node.Memory.TryGetValue(Agent.GetInstanceID(), out object memory))
+                return;
+
+            switch (node.Node)
+            {
+                case Flow flow:
+                    flow.OnFlowEnter(memory, Agent);
+                    break;
+                case Action action:
+                    action.OnNodeEnter(memory, Agent);
+                    break;
+                default:
+                    return;
+            }
+        }
+
+        private void StepOut(ExecutableNode node)
+        {
+            if (!node.Memory.TryGetValue(Agent.GetInstanceID(), out object memory))

```

Comment:
private void StepOut(ExecutableNode node)
{
    UnityEngine.Debug.Log($"Node: {node}");
    UnityEngine.Debug.Log($"node.Memory: {node.Memory}");
    UnityEngine.Debug.Log($"Agent: {Agent}");
    UnityEngine.Debug.Log($"Agent.GetInstanceID(): {Agent.GetInstanceID()}");

    if (!node.Memory.TryGetValue(Agent.GetInstanceID(), out object memory))
        return;

    // [...]
}

Results in:
![image](https://github.com/acdamiani/schema/assets/20422129/9bf4a7f6-0ad6-46fb-9981-0b795b918f61)